File: src/io/xol/chunkstories/server/net/ServerAnnouncerThread.java
package io.xol.chunkstories.server.net;

import java.util.Random;
import java.util.concurrent.atomic.AtomicBoolean;

import io.xol.chunkstories.VersionInfo;
import io.xol.chunkstories.server.Server;
import io.xol.engine.net.HttpRequests;

//(c) 2015-2017 XolioWare Interactive
// http://chunkstories.xyz
// http://xol.io

public class ServerAnnouncerThread extends Thread
{
	AtomicBoolean run = new AtomicBoolean(true);

	int lolcode = 0;
	public long updatedelay = 0;

	public String srv_name;
	public String srv_desc;

	Server server;
	
	public ServerAnnouncerThread(Server server)
	{
		this.server = server;
		
		lolcode = server.getServerConfig().getIntProp("lolcode", "0");
		if (lolcode == 0L)
		{
			// System.out.println("lolcode = 0");
			Random rnd = new Random();
			lolcode = rnd.nextInt(100000);
			server.getServerConfig().setInteger("lolcode", lolcode);
		}
		updatedelay = Long.parseLong(server.getServerConfig().getProp("update-delay", "10000"));
		String hostname = HttpRequests.sendPost("http://chunkstories.xyz/api/sayMyName.php?host=1", "");
		srv_name = server.getServerConfig().getProp("server-name", "unnamedserver@" + hostname);
		srv_desc = server.getServerConfig().getProp("server-desc", "Default description.");
		setName("Multiverse thread");
	}

	public void stopAnnouncer()
	{
		run.set(false);
	}

	@Override
	public void run()
	{
		try
		{
			String ip = HttpRequests.sendPost("http://chunkstories.xyz/api/sayMyName.php?ip=1", "");
			while (run.get())
			{
				// System.out.println("Updating server data on Multiverse.");
				if (server.getServerConfig().getProp("enable-multiverse", "false").equals("true"))
				{
					HttpRequests.sendPost("http://chunkstories.xyz/api/serverAnnounce.php", "srvname=" + srv_name + "&desc=" + srv_desc + "&ip=" + ip + "&mu=" + server.getHandler().getMaxClients() + "&u="
							+ server.getHandler().getNumberOfAuthentificatedClients() + "&n=0&w=default&p=1&v=" + VersionInfo.version + "&lolcode=" + lolcode);
					sleep(updatedelay);
				}
				else
					sleep(6000);
			}
		}
		catch (Exception e)
		{
			server.getLogger().error("An unexpected error happened during multiverse stuff. More info below.");
			e.printStackTrace();
		}
	}
}

File: server/src/main/java/xyz/chunkstories/server/net/announcer/ServerAnnouncerThread.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.server.net.announcer

import xyz.chunkstories.api.util.configuration.Configuration
import xyz.chunkstories.net.http.SimplePostRequest
import xyz.chunkstories.net.http.SimpleWebRequest
import xyz.chunkstories.server.DedicatedServer
import xyz.chunkstories.server.DedicatedServerOptions
import xyz.chunkstories.util.VersionInfo

import java.net.Inet4Address
import java.util.Random
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Small background thread that is tasked with putting and keeping up to date the server's entry in the global list
 * TODO: this is SERIOUSLY OLD junk (circa 2015!)
 */
class ServerAnnouncerThread(private val server: DedicatedServer) : Thread() {
    private val run = AtomicBoolean(true)

    private var lolcode: Int = 0 // TODO this is nowhere enough rng to be secure

    init {

        lolcode = server.config.getIntValue(DedicatedServerOptions.announcerLolcode)
        if (lolcode.toLong() == 0L) {
            val rnd = Random()
            lolcode = rnd.nextInt(Integer.MAX_VALUE)
            val option = server.config.get<Configuration.OptionInt>(DedicatedServerOptions.announcerLolcode)
            option!!.trySetting(lolcode)
        }

        name = "Server list announcer thread"
    }

    fun stopAnnouncer() {
        run.set(false)
    }

    override fun run() {
        try {
            val internalIp = Inet4Address.getLocalHost().hostAddress
            val externalIp = SimpleWebRequest("https://chunkstories.xyz/api/sayMyName.php?ip=1").waitForResult()

            while (run.get()) {
                if (server.config.getBooleanValue(DedicatedServerOptions.announcerEnable)) {

                    val serverName = server.config.getValue(DedicatedServerOptions.serverName)
                    val serverDescription = server.config.getValue(DedicatedServerOptions.serverDescription)

                    val postData = ("srvname=$serverName&desc=$serverDescription&ip=$externalIp&iip=$internalIp&mu=" + server.connectionsManager
                            .maxClients + "&u=" + server.connectionsManager.playersNumber + "&n=0&w=default&p=1&v=" + VersionInfo.versionJson.version
                            + "&lolcode=" + lolcode)

                    SimplePostRequest("https://chunkstories.xyz/api/serverAnnounce.php", postData)

                    sleep(server.config.getIntValue(DedicatedServerOptions.announcerDutyCycle).toLong())
                } else
                    sleep(6000)
            }
        } catch (e: Exception) {
            server.logger.error("An unexpected error happened during multiverse stuff. More info below.")
            e.printStackTrace()
        }

    }
}
